### PR TITLE
Improve handling of rule configmaps

### LINKED ISSUE #5085: Improve handling of rule configmaps
**What is missing?**
Improved handling of rule configmaps in a grow-only mode. 

**Why do we need it?**
When reaching the size-limit of the first configmap and then falling below it again, the operator updates the STS and recreates the pods to go back to 1 configmap. This introduces disturbance - especially if the number of rules is around the limit, therefore re-triggering this behaviour multiple times.

I would propose just leaving the new ConfigMap empty in this case and not removing it, therefore reducing the impact on the running Prometheus.

### PR DESCRIPTION
**Pull request by mikechengwei**
## Description

Close #5085 .

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Improve handling of rule configmaps.
```

### CODE REVIEW

**Review by JoaoBraveCoding — COMMENTED** [review]
IIUC the issue, I don't think the goal would be to expose something like the field `PreProvisionRuleConfigMapNumber` but rather hide that detail from users, especially because it's not clear to users how many ConfigMaps they will need. 
The operator should instead provision a set of extra ConfigMaps (2 or 3) that it would mount on the statefulset and use them to spread the configuration if need be. 
Furthermore I feel that the operator should still maintain the ability to provision more ConfigMaps if the configuration grows past the point where the initial ConfigMaps are not enough anymore 
I'll let one of the maintainers correct me if I misunderstood it

**Comment by simonpasquier** [comment]
I confirm what @JoaoBraveCoding said: there's no need to expose the number of pre-provisioned configmaps to the end-user IMHO. As I hinted in the original issue, we might be able to use [optional configmaps](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#optional-configmaps) to always mount 3 configmaps (3 is an arbitrary number, my hunch is that most deployments won't need more) even though the operator needs only to create one configmap initially.
When more than 3 configmaps are needed, the operator will trigger a statefulset rollout but I think that it's ok. And to avoid subsequent rollout if the number of configmaps shrinks again, I'd record the max number of configmaps in a statefulset's annotation. The operator would use the annotation to know how many volume mounts should be provisioned. Does it make sense?

**Comment by mikechengwei** [comment]
Ok let me think about it.

**Comment by mikechengwei** [comment]
@simonpasquier @JoaoBraveCoding PLTA again.

File: pkg/prometheus/rules_test.go
Comment (JoaoBraveCoding) [inline pkg/prometheus/rules_test.go:133]: forgot to update the error message
```suggestion
		t.Fatalf("expected three ConfigMaps but got %v", len(configMaps))
```

File: pkg/prometheus/rules_test.go
Comment (JoaoBraveCoding) [inline pkg/prometheus/rules_test.go:163]: Forgot to update the error message. But I fee like this test should be updated to continue testing the code path where configuration exceeds the 3 ConfigMaps

**Review by JoaoBraveCoding — COMMENTED** [review]
Good work! But I feel like it's still not there, please take a look at the comment [1] where I provide a bit more detail. 

Disclaimer: I'm not a maintainer so if you have a different opinion feel free to wait for the review of one :) 

[1] https://github.com/prometheus-operator/prometheus-operator/pull/5159#discussion_r1029204404

**Comment by mikechengwei** [comment]
@JoaoBraveCoding plz help me review it again.

File: pkg/k8sutil/k8sutil.go
Comment (JoaoBraveCoding) [inline pkg/k8sutil/k8sutil.go:389]: Comment seems to be outdated
```suggestion
// DiffRulerConfigMap preProvisionRuleConfigMapNumber can't less than newConfigMaps number.
```

File: pkg/k8sutil/k8sutil.go
Comment (JoaoBraveCoding) [inline pkg/k8sutil/k8sutil.go:412]: I'm not sure about this, but here couldn't we use [1] instead of maintaining this code. @simonpasquier might know what's best to do in this specific case.

[1] https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate

File: pkg/prometheus/rules.go
Comment (JoaoBraveCoding) [inline pkg/prometheus/rules.go:114]: Don't think this comment is necessary

File: pkg/prometheus/rules.go
Comment (JoaoBraveCoding) [inline pkg/prometheus/rules.go:134]: This comment has to be updated, it's important because it's the part of the code where we will return the extra empty configMaps that will be mounted optionally. This part of the code will not be obvious if the person reading the code does not know about #5085

File: pkg/prometheus/rules_test.go
Comment (JoaoBraveCoding) [inline pkg/prometheus/rules_test.go:134]: Sorry, but since we reverted our change to `makeRulesConfigMaps` this should be reverted as well
```suggestion
		t.Fatalf("expected one ConfigMaps but got %v", len(configMaps))
```

File: pkg/prometheus/rules_test.go
Comment (JoaoBraveCoding) [inline pkg/prometheus/rules_test.go:164]: Same as previous comment
```suggestion
		t.Fatalf("expected rule files to be split up into two ConfigMaps, but got '%v' instead", len(configMaps))
```

File: pkg/prometheus/rules_test.go
Comment (JoaoBraveCoding) [inline pkg/prometheus/rules_test.go:181]: Since here we are testing only the function `makeRulesConfigMaps` I don't think adding this test makes sense as it's testing the same behaviour as `shouldSplitUpLargeSmallIntoTwo`. Instead, it might be better to test `createOrUpdateRuleConfigMaps` since that's the function that we actually changed the behaviour

File: pkg/prometheus/statefulset.go
Comment (JoaoBraveCoding) [inline pkg/prometheus/statefulset.go:518]:
```suggestion
		optional := pointer.Bool(true)
		if index == 0 {
			optional = pointer.Bool(false)
		} 
```

**Review by JoaoBraveCoding — COMMENTED** [review]
I think it's much closer to the proposition @simonpasquier left! Just a few more comments but I'm pretty happy with the direction! Great work and thank you for your work! 💪 

PS: the comments I left to `prometheus/(rules.go/statefulset.go)` should also apply to `thanos/(rules.go/statefulset.go)`

File: pkg/k8sutil/k8sutil.go
Comment (mikechengwei) [reply pkg/k8sutil/k8sutil.go:412]: I don't think it's good to use this method? If yes, why not all create resources using this package.

**Comment by mikechengwei** [comment]
@simonpasquier  can u help me review it again?

File: pkg/prometheus/rules.go
Comment (JoaoBraveCoding) [inline pkg/prometheus/rules.go:133]:
```suggestion
	// supple configmap volumeMount number, at least `default` number of rule configmaps mount.
	// this is to prevent the StatefulSet Pods from being frequently redeployed when the size of the
	// configuration can easily fluctuate between the maximum size of a single ConfigMaps or not.
```

**Comment by mikechengwei** [comment]
> Just a minor comment but apart from that it looks great! Great job, and thanks for addressing my comments even though I'm not a maintainer of the project 👍

Thanks for your suggestions ,happy cooperation.

**Comment by JoaoBraveCoding** [comment]
@mikechengwei can you fix the failing CI

**Comment by mikechengwei** [comment]
> @mikechengwei can you fix the failing CI



> @mikechengwei can you fix the failing CI

ok.

**Review by JoaoBraveCoding — APPROVED** [review]
lgtm 👍 

Thanks for all the work, hope we can merge this in the future

**Comment by mikechengwei** [comment]
> lgtm 👍
> 
> Thanks for all the work, hope we can merge this in the future

I thought so too, but it seems that no one in the community has reviewed the PR.

**Comment by PhilipGough** [comment]
@mikechengwei - sorry, I missed this one but will get it reviewed ASAP

**Comment by mikechengwei** [comment]
> sorry, I missed this one but will get it reviewed ASAP

Is there any new progress? @PhilipGough

**Pull request closed without merging**
